src/benchq/algorithms/utils/convex_optimization.py

```diff
@@ -0,0 +1,117 @@
+            in terms of Chebyshev functions.
+        degree (int): degree of opproximating polynomial.
+    """
+
+    def chebychev_expansion(x, coeff):
+        """Function that approximates a desired one using Chebyshev polynomials."""
+        return [
+            sum(
+                coeff[k] * np.cos((2 * k + 1) * np.arccos(xi))
+                for k in range((degree - 1) // 2)
+            )
+            for xi in x
+        ]
+
+    def exact_func(x):
+        """Function we are trying to approximate."""
+        return (1 - delta) * gamma_prime * x
+
+    grid_points = 500
+    mesh_size = 10000
+    constraints_const = max(0.999, 1 - 0.1 * delta)
+    # roots of Chebyshev's polynomial
+    cheb_roots = [-np.cos(np.pi * j / (grid_points - 1)) for j in range(grid_points)]
+    # non-zero roots to force interval constraints
+    cheb_roots_reduced = [
+        root for root in cheb_roots if root >= 0 and root <= 1 / gamma_prime
+    ]
+
+    degree = 3
+    epsilon = 1.0
+    iter_step = 0
+    while True:
+        coeff_k = cp.Variable((degree - 1) // 2)  # variables to optimize for
+        approximation = cp.hstack(
+            [
+                cp.sum(
+                    [
+                        coeff_k[k] * np.cos((2 * k + 1) * np.arccos(cheb_root))
+                        for k in range((degree - 1) // 2)
+                    ]
+                )
+                for cheb_root in cheb_roots_reduced
+            ]
+        )
+        # vectorizing constraints
+        norm_constraint = []
+        for cheb_root in cheb_roots:
+            single = cp.sum(
+                [
+                    coeff_k[k] * np.cos((2 * k + 1) * np.arccos(cheb_root))
+                    for k in range((degree - 1) // 2)
+                ]
+            )
+            norm_constraint.append(cp.norm(single, "inf") <= constraints_const)
+
+        target_val = cp.hstack(
+            [exact_func(cheb_root) for cheb_root in cheb_roots_reduced]
+        )
+        objective = cp.Minimize(cp.norm(target_val - approximation, "inf"))
+        problem = cp.Problem(objective, norm_constraint)
+        problem.solve()
+
+        if problem.status == "optimal":
+            approximation_coeffs = coeff_k.value
+            x = np.linspace(-1 / gamma_prime, 1 / gamma_prime, mesh_size)
+            y_act = np.array([exact_func(xi) for xi in x])
+            y_appr = np.array(chebychev_expansion(x, approximation_coeffs))
+            epsilon = float(np.linalg.norm(y_act - y_appr, ord=np.inf))
+            if epsilon <= error:
+                print("Accuracy is reached")
+                return approximation_coeffs, epsilon, degree
+            elif iter_step > maxiter:
+                print("Max iterations is reached")
+                return approximation_coeffs, epsilon, degree
+        else:
+            print("Problem is not feasible")
+            return np.zeros((1,)), epsilon, degree
```

Comment:
I think that here we'll want to raise an exception rather than return zeroes. This will prevent users from unintentionally using coefficients that don't actually meet the desired error threshold.